### PR TITLE
formula: don't expand unused optional dependencies

### PR DESCRIPTION
**Pull request by alyssais**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

—--

This properly addresses the issue reported in Homebrew/homebrew-core#7826, so `Formula#runtime_dependencies` no longer attempts to expand optional dependencies that aren't being used in the current build.

As such, it reverts #1665, and so unreverts #1592.

### CODE REVIEW

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -1505,7 +1505,12 @@ def recursive_requirements(&block)
   # Returns a list of Dependency objects that are required at runtime.
   # @private
   def runtime_dependencies
-    recursive_dependencies.reject(&:build?)
+    recursive_dependencies do |_dependent, dependency|
+      Dependency.prune if dependency.build?
+      if dependency.optional? || dependency.recommended?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: `if build.without?(dependency) && (dependency.optional? || dependency.recommended?)` may be easier to follow. Could also consider having `optional_dependency = dependency.optional? || dependency.recommended?` before that, too.

**Comment by MikeMcQuaid** [comment]
@alyssais when you consider this good to 🚢, can you provide a wee test plan I can run locally to reproduce the issue this is fixing? Thanks!

**Comment by alyssais** [comment]
@MikeMcQuaid I reproduced the issue by running `Formula['ffmpeg'].runtime_dependencies` in `brew irb`, and seeing if it produced an exception or not. I imagine you'll want a more realistic test scenario, so I'll see if I can come up with one…

File: Library/Homebrew/formula.rb (hunk 1 above)
Comment (alyssais) [inline Library/Homebrew/formula.rb]: How's that @MikeMcQuaid?

**Comment by alyssais** [comment]
Going to force push because the history has ended up a bit unclear here…

**Comment by alyssais** [comment]
Okay, there are two things to check here (I assume building from source):

1. Does `brew install influxdb` record `go` as a runtime dependency in `influxdb`'s install receipt? The correct behaviour is not to.
2. Does `brew install https://raw.githubusercontent.com/Homebrew/homebrew-core/636e5b8ee684c4885f0a296cf8bf27f6faa9eba5/Formula/ffmpeg.rb
&& bin/brew upgrade ffmpeg` succeed, or does the upgrade fail with a "No available formula with the name "homebrew/science/opencv"" error? The correct behaviour is for it to succeed.

Currently, only _2_ should succeed. As of 3f72482, only _1_ should succeed. With this PR as a whole, both _1_ and _2_ should succeed.

**Comment by MikeMcQuaid** [comment]
Thanks again @alyssais!

**Pull request merged**